Bluetooth: Host: Add generic pairing query callback

demantmrrv opened the pull request:
Similar to pairing_confirm this callback is called each
time a peer requests pairing, but for all types of
pairings, except SSP. The pairing req/rsp information is
passed as a parameter so the application can decide
wheter to accept or reject the pairing.

Fixes: #21036

Signed-off-by: Martin Rieva <mrrv@demant.com>

demantmrrv committed:
Bluetooth: Host: Add generic pairing query callback

Similar to pairing_confirm this callback is called each
time a peer requests pairing, but for all types of
pairings, except SSP. The pairing req/rsp information is
passed as a parameter so the application can decide
wheter to accept or reject the pairing.

Fixes: #21036

Signed-off-by: Martin Rieva <mrrv@demant.com>

joerchan commented on include/bluetooth/conn.h:
```diff
@@ -711,8 +711,61 @@ struct bt_conn_oob_info {
+
 /** Authenticated pairing callback structure */
 struct bt_conn_auth_cb {
+#if defined(CONFIG_BT_SMP_PAIRING_ALLOW)
```
Can you predefine this in doxygen so that the documentation for this will be included?

Look for `PREDEFINED` is `doc/zephyr.doxyfile.in`

demantmrrv replied:
I'm a bit in doubt on that one. I've added Doxygen comments on the structures themselves, but looking at `doc/zephyr/doxyfile.in` it seems that only major flags are present there. Are you sure it should be added there?

joerchan replied:
@demantmrrv Yes, I am sure. It needs to be added there so that it will not excluded from the online documentation.
All #ifdes that are in public header files are enabled there so that all features are included in the documentation.

demantmrrv replied:
Oh, okay, I wasn't aware of that. I'll add it right away.

demantmrrv commented on include/bluetooth/conn.h:
```diff
@@ -711,8 +711,61 @@ struct bt_conn_oob_info {
+	 *  This callback may be unregistered in which case pairing continues
+	 *  as if the Kconfig flag was not set.
+	 *
+	 * Currently BT SSP is not supported.
```
Yes, hence the "Currently BT SSP is not supported." line. Should it be more explicit?

joerchan replied:
I wanted to change to that line instead. SSP has not been used in the header so would make sense to define what it means. And otherwise BR/EDR is more clear, as SSP is only relevant for BR/EDR. BT is never used as an abbreviation, we always spell use either Bluetooth LE or BR/EDR.

carlescufi commented:
@demantmrrv please rebase correctly so that you don't include unrelated commits

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -2111,6 +2138,27 @@ static u8_t send_pairing_rsp(struct bt_smp *smp)
 #endif /* CONFIG_BT_PERIPHERAL */
 
+static u8_t smp_pairing_allow_query(
+	struct bt_conn *con,
```
Use `conn` for consistency with existing code. Also, move it to the first line since you should only be splitting when you otherwise go beyond 80 characters.

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -404,6 +404,33 @@ static enum bt_security_err auth_err_get(u8_t smp_err)
 	}
 }
 
+#if defined(BT_SMP_PAIRING_ALLOW)
```
It's `CONFIG_BT_SMP_PAIRING_ALLOW`, not `BT_SMP_PAIRING_ALLOW`, i.e. it seems you never tested or even compiled this piece of code.

demantmrrv replied:
OMG, good catch. The flag name was changed last minute, but the same flag was used in the tests, so they did run, but the code was never called after that. Thanks!

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -2111,6 +2138,27 @@ static u8_t send_pairing_rsp(struct bt_smp *smp)
+	struct bt_conn *con,
+	struct bt_smp_pairing *pairing)
+{
+#if defined(BT_SMP_PAIRING_ALLOW)
```
`CONFIG_BT_SMP_PAIRING_ALLOW`. What worries me a bit is that this indicates that you never tested the code.

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -2810,6 +2858,16 @@ static u8_t smp_pairing_req(struct bt_smp *smp, struct net_buf *buf)
 #if defined(CONFIG_BT_SMP_SC_PAIR_ONLY)
 		return BT_SMP_ERR_AUTH_REQUIREMENTS;
 #else
+		if (IS_ENABLED(BT_SMP_PAIRING_ALLOW)) {
```
`CONFIG_BT_SMP_PAIRING_ALLOW`

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -2826,6 +2884,15 @@ static u8_t smp_pairing_req(struct bt_smp *smp, struct net_buf *buf)
 		return BT_SMP_ERR_ENC_KEY_SIZE;
 	}
 
+	if (IS_ENABLED(BT_SMP_PAIRING_ALLOW)) {
```
`CONFIG_BT_SMP_PAIRING_ALLOW`

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -2992,6 +3059,16 @@ static u8_t smp_pairing_rsp(struct bt_smp *smp, struct net_buf *buf)
 #if defined(CONFIG_BT_SMP_SC_PAIR_ONLY)
 		return BT_SMP_ERR_AUTH_REQUIREMENTS;
 #else
+		if (IS_ENABLED(BT_SMP_PAIRING_ALLOW)) {
```
`CONFIG_BT_SMP_PAIRING_ALLOW`

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3011,6 +3088,15 @@ static u8_t smp_pairing_rsp(struct bt_smp *smp, struct net_buf *buf)
 	smp->local_dist &= SEND_KEYS_SC;
 	smp->remote_dist &= RECV_KEYS_SC;
 
+	if (IS_ENABLED(BT_SMP_PAIRING_ALLOW)) {
```
`CONFIG_BT_SMP_PAIRING_ALLOW`

jhedberg commented on include/bluetooth/conn.h:
```diff
@@ -711,8 +711,78 @@ struct bt_conn_oob_info {
+	/** IO Capability, Core Spec. Vol 3, Part H, 3.5.1, Table 3.4 */
+	u8_t io_capability;
+	/** OOB data flag, Core Spec. Vol 3, Part H, 3.5.1, Table 3.5 */
+	u8_t oob_data_flag;
```
For improved readability could you add an empty line after each variable. That makes it easier to see where the documentation for the previous variable ends and a new variable definition begins.

jhedberg commented on include/bluetooth/conn.h:
```diff
@@ -711,8 +711,78 @@ struct bt_conn_oob_info {
+	     sc:1,
+	     keypress:1,
+	     ct2:1,
+	     rfu:2;
```
This is unportable since you're assigning to it the raw u8_t value from the protocol. The C standard doesn't define how bitfields are ordered in memory (it's left to be implementation-defined). Usually little- and big-endian architectures do this differently.

We've generally steered away from bitfields when it comes to protocol values, and only used them for internal variables where compactness was desired. Instead we use helper macros to encode and decode to the protocol format.

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -320,6 +320,16 @@ config BT_SIGNING
 	  authenticated data in an unencrypted connection.
 
+config BT_SMP_PAIRING_ALLOW
+	bool "Allow pairing to proceed"
```
Do you think the term "validate" might be more intuitive than "allow" here? Since essentially the callback is validating the pairing parameters against what it considers acceptable. I'd also add "param" or "req" somewhere in there (also in the code) to make it clear that it's validating the parameters of the pairing request rather than the pairing itself.

demantmrrv replied:
We did have some internal talks about naming and changed to allow, as it is free for the application to include these during processing or not. Also, it is planned to extend this to include security_req also (with AuthReq only), keeping the same flag. So, yes, naming was considered.

jhedberg replied:
The reason "allow" bothers me is that it describes one of two possible results of the action, not the action itself. The action is something like "verify", "confirm" or "validate", and the result can be either "allow" or "deny". Calling the option PAIRING_DENY would be equally valid using the current reasoning, which I hope you agree would look strange. So I'd rather have this refer to the action and not the result.

Note: the reason I'm being slightly pedantic about this is that this becomes a public API, and those are always painful to change/correct after the fact. I'd rather we get this right from the beginning.

demantmrrv replied:
Well, "confirm" is already taken (although biased towards "allow", as well are "verify" and "validate") and I wanted to avoid confusion with it.
So, lifting this to a holistic synonym of taking a decision we also have "resolve", "resolution", "conclude", "regulate", "react", "consent", "authorize", "authenticate" and probably others too.

jhedberg replied:
Would prefixing this with "pre" make it any better? E.g. pre_confirm?

joerchan replied:
Maybe we could include where in the pairing stage this is to make it more clear?
pairing_features_confirm, or some other word for confirm? pairing_feature_exchange_confirm?
Maybe accept could work in place of confirm?

demantmrrv replied:
Well, "confirm" in its entirety does not adhere to the argument of an action based and unbiased naming convention.
Maybe a combination as @joerchan suggests, such as "pairing_properties_audit", "pairing_features_filter" or the like - maybe with "peer" inserted, if not redundant.

demantmrrv replied:
@jhedberg let's resolve this naming issue with a name that suits you best.

jhedberg reviewed (changes requested):
It seems most of this code was never being tested (or even compiled) since you were using the wrong pre-processor variable for Kconfig.

demantmrrv commented:
Built and tested OK without Kconfig flag and with Kconfig flag with app returning either zero or non-zero value in callback.

jhedberg commented:
@demantmrrv we had a short chat with @aescolar and @joerchan with the following concensus:

 - Rename `bt_conn_pairing_info` to `bt_conn_pairing_feat` since this "information" is called "features" in the spec (as in "Pairing Feature Exchange"). The parameter name would then just be `feat`.
 - Rename the callback to `pairing_accept` (it's for asking the application if the given pairing features can be accepted)
 - Rename the Kconfig variable to `BT_SMP_USER_PAIRING_ACCEPT` (having "user" in there is clearer in that it's about enabling something extra for the user to accept, rather than looking like some global boolean allow/deny option)

demantmrrv commented:
@jhedberg About the latter bullet; the "user" ie. person, should not be informed about this as it is a synchronous call. If "user" indicates the application, no worries then.

jhedberg commented:
> @jhedberg About the latter bullet; the "user" ie. person, should not be informed about this as it is a synchronous call. If "user" indicates the application, no worries then.

It's more of a reference to the "user" of the API, i.e. the application. One of the suggestions from @joerchan was actually APP. We can still reconsider this. I suggested CUSTOM since the stack has some existing built-in checks and this is about an additional one to be done by the user of the API.

joerchan commented:
> It's more of a reference to the "user" of the API, i.e. the application. One of the suggestions from @joerchan was actually APP. We can still reconsider this. I suggested CUSTOM since the stack has some existing built-in checks and this is about an additional one to be done by the user of the API.

So usually (with nRF5 SDK) we have referred to these things as "Application control of <feature>" when give control over something that would normally be stack controller. Examples of this are data length and phy.
Which is why I initially suggested APP, but I'd prefer to be consistent with other kconfigs as well.
I remember someone started to use "USER" for this, for example `/** Register user callback for HCI Vendor-Specific Events` and it's kconfig `CONFIG_BT_HCI_VS_EVT_USER`.
It seems it's not widely used, only other instance would be LVGL_MEM_POOL_USER that I could find.

So personally I would prefer APP because as @demantmrrv pointed out we want to distinguish in pairing from the "User input" i.e a person typing in the passkey and the application without user involvement.

demantmrrv commented:
Based on @joerchan, @jhedberg and @aescolar's comments I have pushed the latest changes now. RFC.

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -2111,6 +2138,26 @@ static u8_t send_pairing_rsp(struct bt_smp *smp)
+{
+#if defined(CONFIG_BT_SMP_APP_PAIRING_ACCEPT)
+	if (bt_auth && bt_auth->pairing_accept) {
+		struct bt_conn_pairing_feat feat = {
```
I'd make this `const struct ...`  here to make it clear that this is not to be modified by the callback, and also make the pointer to the callback use `const`

demantmrrv replied:
Can do that. Thought of it, but it doesn't really matter as long as it is isolated from the packet data.

jhedberg replied:
It might give the compiler more opportunities for optimisation though, right? At least I'd have thought for this reason it to be good practice to declare everything const that can be declared const.

jhedberg replied:
Anyway, I don't really have a super strong opinion on this, however as the callback is a public API we should avoid changing the types after it's merged upstream.

demantmrrv replied:
Shouldn't we change `struct bt_conn *conn` to `const struct bt_conn *const conn` as well? Then we use the same practice.

jhedberg replied:
`bt_conn` is a little bit of a different story. It's an opaque type, i.e. the application has no visibility into its implementation and cannot do direct access to its members. It's also valid for the application to want to call some API which takes a bt_conn pointer from within the callback, in which case things would break (create a compiler error) since no bt_conn-taking APIs use const.

demantmrrv replied:
Alright, just added const.

joerchan commented on subsys/bluetooth/host/smp.c:
```diff
@@ -2992,6 +3058,16 @@ static u8_t smp_pairing_rsp(struct bt_smp *smp, struct net_buf *buf)
+		if (IS_ENABLED(CONFIG_BT_SMP_APP_PAIRING_ACCEPT)) {
+			u8_t err;
+
+			err = smp_pairing_accept_query(smp->chan.chan.conn,
```
This should be rsp

joerchan commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3011,6 +3087,15 @@ static u8_t smp_pairing_rsp(struct bt_smp *smp, struct net_buf *buf)
+	if (IS_ENABLED(CONFIG_BT_SMP_APP_PAIRING_ACCEPT)) {
+		u8_t err;
+
+		err = smp_pairing_accept_query(smp->chan.chan.conn, req);
```
this should be rsp

joerchan committed:
Bluetooth: shell: Add support for pairing accept callback

Add pairing support callback to print remote pairing features when this
option is enabled.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>
